aiohttp/web_urldispatcher.py

```diff
@@ -998,20 +1008,41 @@ def __init__(self) -> None:
         self._named_resources: Dict[str, AbstractResource] = {}
         self._resource_index: dict[str, list[AbstractResource]] = {}
         self._matched_sub_app_resources: List[MatchedSubAppResource] = []
+        self._hyperdb: Optional[hyperscan.Database] = None  # type: ignore[no-any-unimported]
+        self._plain_resources: dict[str, PlainResource] = {}
+        self._prefix_resources: dict[str, list[PrefixResource]] = {}
+        self._has_variable_resources = True
+
+    def _on_match(
+        self, id_: int, from_: int, to: int, flags: int, found: list[int]
+    ) -> Optional[bool]:
+        found.append(id_)
+        return None
 
     async def resolve(self, request: Request) -> UrlMappingMatchInfo:
-        resource_index = self._resource_index
-        allowed_methods: Set[str] = set()
+        allowed_methods: set[str] = set()
+        path = request.rel_url.path_safe
+
+        # plain resource lookup
+        if (plain_resource := self._plain_resources.get(path)) is not None:
+            match_dict, allowed = await plain_resource.resolve(request)
+            if match_dict is not None:
+                return match_dict
+            else:
+                allowed_methods |= allowed
+
+        # prefix resource lookup
+        url_part = path
+        prefix_resources = self._prefix_resources
 
         # Walk the url parts looking for candidates. We walk the url backwards
         # to ensure the most explicit match is found first. If there are multiple
         # candidates for a given url part because there are multiple resources
         # registered for the same canonical path, we resolve them in a linear
         # fashion to ensure registration order is respected.
-        url_part = request.rel_url.path_safe
         while url_part:
```

Comment:
If we're back to this approach, I'm wondering if it worth merging plain and prefix resources again. Might be a tiny bit slower when a plain resource matches and a tiny bit faster in other cases. But, would also reduce the amount of code to maintain.

Reply:
I'm still working on the PR, who knows how it will look at the end.
The existing index table is suboptimal for prefix resources.
Assume we have `/app` prefix subapp.
For handling path like `/app/path/inside/subapp` the code tries `/app/path/inside/subapp`, `/app/path/inside`, `/app/path`, and the only last `/app` test succeeds.
I would like to avoid both linear scan *and* walking up to the path for prefix resources.
The optimization target is good results for `test_resolve_gitapi_subapps` and `test_resolve_prefix_resources_many_prefix_many_plain` benchmark, the later was added by https://github.com/aio-libs/aiohttp/pull/9935
If you imagine other common cases for prefixed resources please share it. We can add another benchmarks and optimize the code to fit them as well.